### DISCUSSION TITLE
Glass BSDF appears black with PTracer Integrator

### DISCUSSION THREAD

**Discussion by h-OUS-e** [mitsuba-renderer/mitsuba3#549] [Q&A] [answered]
I was wondering why does a transparent material appear black when rendering using a ptracer integrator. but with path you can see through it, but still not as clear as I hoped.

![image](https://user-images.githubusercontent.com/66432003/220024867-2d5002d1-aef2-443a-9fa8-00770a34c260.png)
![image](https://user-images.githubusercontent.com/66432003/220025071-9b8d9c2b-4149-47a8-9241-92168120f7c1.png)

**Comment by h-OUS-e** [comment, marked as answer]
PS: I was using a dielectric material. My initial guess is that  with the PTracer, it would be hard to find rays that are emitted from the light source, and hit the camera exactly?

To test this, I made the red ball as the only light source.

PTracer with Dielectric Glass Slab
![image](https://user-images.githubusercontent.com/66432003/220026477-0eadaf43-ce30-46f5-b8ff-4d4544e0a6cd.png)

PTracer with RoughDielectric Glass Slab
![image](https://user-images.githubusercontent.com/66432003/220026444-8a3aa045-45f7-4562-ae4a-22c1901ecba4.png)

Path with Dielectric Glass Slab
![image](https://user-images.githubusercontent.com/66432003/220026706-02e10640-8579-4a5c-9f9f-d5fdbf7d1dc5.png)

Path with RoughDielectric Glass Slab
![image](https://user-images.githubusercontent.com/66432003/220026596-aebbbf01-3108-4ebc-aede-049829e251dc.png)

Still curious if we can use the PTracer with regular Dielectric and see through the glass. My initial guess was that we could given the caustic optimization tutorial works.

**Reply by njroussel** [reply]
Hi @h-OUS-e 
 
 You are correct it is "_hard to find rays that are emitted from the light source and hit the camera_" hence it being black with the `ptracer`.

The tutorial has a slightly different setup, the sensor is positioned between the glass and the receiving plane, and is looking at the receiving plane. Effectively, the ptracer can "connect" to the sensor because the last bounce it has is with the receiving plane (diffuse material) and not the lend (dielectric material). This is described in [this section](https://mitsuba.readthedocs.io/en/latest/src/inverse_rendering/caustics_optimization.html#Assembling-the-scene) of the tutorial.

**Reply by h-OUS-e** [reply]
So for the ptracer to pick up the light rays after passing through a dielectric material, they have to scatter somehow? Is that why the roughdielectric material works with the ptracer? Does scattering light make it easier to find the rays? I would've imagined it is easy to find light rays coming from an emitter if I am looking at the emitter straight on through a piece of glass, since the light rays would still travel straight into the camera.

**Reply by njroussel** [reply]
> So for the ptracer to pick up the light rays after passing through a dielectric material, they have to scatter somehow? Is that why the roughdielectric material works with the ptracer? 
Yes, exactly.

> Does scattering light make it easier to find the rays?
Generally, yes.

A good way to gain some intuition about this is to think about how the rays are constructed. For the `ptracer`, the ray is spawned from the emitter, then it will go straight into the dielectric, refract in it, and exit it. When exiting, you'll actually want it to "connect" to the sensor. However, the exiting angle will most likely never match the exact angle you need to hit the origin (pinhole) of a `perspective` sensor. Just imagine how unlikely it is for any random ray to pass through some arbitrary pre-defined point.
Some scattering will help because we can usually always "connect" to the sensor, we just need to account for the probability of that specific scattering direction to happen.
